torch/csrc/autograd/variable.cpp

```diff
@@ -154,6 +154,25 @@ AutogradMeta* materialize_autograd_meta(const at::TensorBase& self) {
   return get_autograd_meta(self);
 }
 
+void update_retains_grad(
+    const at::TensorBase& self,
+    const std::shared_ptr<torch::autograd::Node>& fn) {
+  const auto& meta = impl::get_autograd_meta(self);
+  if (!self.retains_grad()) {
+    meta->cpp_hooks_list_ = nullptr;
+    return;
+  }
+  auto idx = meta->retains_grad_;
+  auto new_list = std::make_shared<hooks_list>();
+  new_list->push_back(std::move((*meta->cpp_hooks_list_)[idx]));
+  (*meta->cpp_hooks_list_)[idx] = nullptr;
+  meta->cpp_hooks_list_ = new_list;
+  meta->retains_grad_ = 0;
```

Comment:
nit: small comment here that since this is a new list, 0 is the index of the retains_grad hook